### PR TITLE
[MetalPerformanceShaders] Neural Networks Update to Xcode 11

### PR DESCRIPTION
**Pull request by dalexsoto**
This includes updates from PRs xamarin/xamarin-macios#6932 and xamarin/xamarin-macios#6935, this is a rebase of xamarin/xamarin-macios#7461.

It adds new functionality to the neural network components.
This is still not the complete API for 11.3

* Update src/metalperformanceshaders.cs

Co-Authored-By: Rolf Bjarne Kvinge <rolf@xamarin.com>

* Make changes requested by @rolfbjarne

* Fix binding of default random distribution creation

The parameterless function creates new
default distributions and is not a property.
It is a counterpart of CreateUniform.

* Expose public APIs for MPSCnnConvolutionTransposeNode

These APIs are public and documented at: https://developer.apple.com/documentation/metalperformanceshaders/mpscnnconvolutiontransposenode/2942641-initwithsource?language=objc

I have also tested that they work.

* Reintroduce compat API.

* Fix acronym casing.

* Fix introspection tests.

* Fix xtro.

* One last xtro issue.

* Fix more xtro.

* Another introspection fix.

Co-Authored-By: Frank A. Krueger <fak@praeclarum.org>
Co-authored-by: Rolf Bjarne Kvinge <rolf@xamarin.com>

### CODE REVIEW

File: src/MetalPerformanceShaders/MPSDefs.cs [hunk 1]
```diff
@@ -423,7 +457,8 @@ public enum MPSNNComparisonType : ulong {
 		GreaterOrEqual,
 	}
 
-	[TV (11, 3), Mac (10, 13, 4), iOS (11, 3)]
+	[Introduced (PlatformName.MacCatalyst, 13, 0)]
+	[TV (11,3), Mac (10,13,4), iOS (11,3)]
```
Comment (rolfbjarne) [inline src/MetalPerformanceShaders/MPSDefs.cs:461]: Reduce whitespace noise:

```suggestion
	[TV (11, 3), Mac (10, 13, 4), iOS (11, 3)]
```

File: src/MetalPerformanceShaders/MPSDefs.cs [hunk 2]
```diff
@@ -438,7 +473,8 @@ public enum MPSCnnLossType : uint {
 		//Count, // must always be last, and because of this it will cause breaking changes.
 	}
 
-	[TV (11, 3), Mac (10, 13, 4), iOS (11, 3)]
+	[Introduced (PlatformName.MacCatalyst, 13, 0)]
+	[TV (11,3), Mac (10,13,4), iOS (11,3)]
```
Comment (rolfbjarne) [inline src/MetalPerformanceShaders/MPSDefs.cs:477]:
```suggestion
	[TV (11, 3), Mac (10, 13, 4), iOS (11, 3)]
```

File: src/metalperformanceshaders.cs [hunk 3]
```diff
@@ -5088,6 +5323,151 @@ interface MPSCnnUpsamplingBilinearGradientNode {
 		double ScaleFactorY { get; }
 	}
 
+	interface IMPSNNLossCallback { }
+
+	[Protocol]
+	[BaseType (typeof(NSObject))]
+	[Model (AutoGeneratedName = true)]
```
Comment (spouliot) [inline src/metalperformanceshaders.cs:5330]: why a model here ? it's not a `*Delegate` and we can't provide a directly usable implementation

File: src/metalperformanceshaders.cs [hunk 4]
```diff
@@ -5885,6 +6273,155 @@ interface MPSKeyedUnarchiver : MPSDeviceProvider {
...
+
+		[Export ("reshapeWithShape:")]
+		void Reshape ([BindAs (typeof (nuint []))] NSNumber[] shape);
+	}
+
+	interface IMPSNDArrayAllocator { }
+
+	[Protocol]
+	[BaseType (typeof(NSObject))]
+	[Model (AutoGeneratedName = true)]
```
Comment (spouliot) [inline src/metalperformanceshaders.cs:6329]: why the `[Model]` ?

File: tests/introspection/ApiSelectorTest.cs [hunk 5]
```diff
@@ -52,6 +52,13 @@ protected virtual bool Skip (Type type)
 				if (ca is ModelAttribute)
 					return true;
 			}
+
+			switch (type.FullName) {
+				case "MetalPerformanceShaders.MPSCommandBuffer":
```
Comment (spouliot) [inline tests/introspection/ApiSelectorTest.cs:57]: minor: alignment

File: src/metalperformanceshaders.cs [hunk 6]
```diff
@@ -1775,6 +1784,85 @@ interface MPSTemporaryImage {
...
+		[Export ("initWithDevice:")]
+		[DesignatedInitializer]
+		IntPtr Constructor (IMTLDevice device);
+	}
+
+	interface IMPSHeapProvider {}
+
+	[Protocol]
+	[BaseType (typeof(NSObject))]
+	[Model (AutoGeneratedName = true)]
```
Comment (spouliot) [inline src/metalperformanceshaders.cs:1815]: why ?

File: src/metalperformanceshaders.cs (hunk 3 above)
Comment (dalexsoto) [reply src/metalperformanceshaders.cs:5330]: Good catch

**Comment by dalexsoto** [comment]
@spouliot Good catch! Fixed thanks 👍

**Pull request merged**